File: examples/governance/compliance_domain/lib/compliance_domain/exemption/exemption.rb
require 'hecks/model'

module ComplianceDomain
  class Exemption
    autoload :Lifecycle, "compliance_domain/exemption/lifecycle"

    include Hecks::Model

    attribute :model_id
    attribute :policy_id
    attribute :requirement
    attribute :reason
    attribute :approved_by_id
    attribute :approved_at
    attribute :expires_at
    attribute :scope
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def requested?; status == "requested"; end
    def active?; status == "active"; end
    def revoked?; status == "revoked"; end

    private

    def validate!
      raise ValidationError, "model_id can't be blank" if model_id.nil? || (model_id.respond_to?(:empty?) && model_id.empty?)
      raise ValidationError, "policy_id can't be blank" if policy_id.nil? || (policy_id.respond_to?(:empty?) && policy_id.empty?)
    end
  end
end

File: examples/governance_static_go/domain/exemption.go
package domain

import (
	"time"
	"github.com/google/uuid"
)

type Exemption struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	ModelId string `json:"model_id"`
	PolicyId string `json:"policy_id"`
	Requirement string `json:"requirement"`
	Reason string `json:"reason"`
	ApprovedById string `json:"approved_by_id"`
	ApprovedAt time.Time `json:"approved_at"`
	ExpiresAt time.Time `json:"expires_at"`
	Scope string `json:"scope"`
	Status string `json:"status"`
}

func NewExemption(modelId string, policyId string, requirement string, reason string, approvedById string, approvedAt time.Time, expiresAt time.Time, scope string, status string) *Exemption {
	a := &Exemption{
		ID:        uuid.New().String(),
		ModelId: modelId,
		PolicyId: policyId,
		Requirement: requirement,
		Reason: reason,
		ApprovedById: approvedById,
		ApprovedAt: approvedAt,
		ExpiresAt: expiresAt,
		Scope: scope,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *Exemption) Validate() error {
	if a.ModelId == "" {
		return &ValidationError{Field: "model_id", Message: "model_id can't be blank"}
	}
	if a.PolicyId == "" {
		return &ValidationError{Field: "policy_id", Message: "policy_id can't be blank"}
	}
	return nil
}
